src/OrchardCore.Modules/OrchardCore.AdminTrees/Services/AdminTreeNavigationProvidersCoordinator.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using OrchardCore.AdminTrees.Models;
+using OrchardCore.Navigation;
+using YesSql;
+
+namespace OrchardCore.AdminTrees.Services
+{
+    // Retrieves all instances of "IAdminNodeNavigationBuilder"
+    // Those are classes that add new "AdminNodes" to a "NavigationBuilder" using custom logic specific to the module that register them.
+    // This class handles their inclusion on the admin menu.
+    // This class is itself one more INavigationProvider so it can be called from this module's AdminMenu.cs
+    public class AdminTreeNavigationProvidersCoordinator : INavigationProvider
+    {
+        private readonly IEnumerable<IAdminNodeNavigationBuilder> _nodeBuilders;
+        private readonly ISession _session;
+        private readonly ILogger Logger;
+
+        public AdminTreeNavigationProvidersCoordinator(
+            ISession session,
+            IEnumerable<IAdminNodeNavigationBuilder> nodeBuilders,
+            ILogger<AdminTreeNavigationProvidersCoordinator> logger)
+        {
+            _session = session;
+            _nodeBuilders = nodeBuilders;
+            Logger = logger;
+        }
+
+        // We only add them if the caller uses the string "admintree").
+        // todo: use a public constant for the string
+        public async Task BuildNavigationAsync(string name, NavigationBuilder builder)
+        {
+            if (!String.Equals(name, "admintree", StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            var trees = (await _session.Query<AdminTree>().ListAsync())
```

Comment:
So apparently you store as many admin trees. But I think you just need a single document in the database. And have it cached in a custom service.

Reply:
Done.